Questions/collections list reorganization + no questions empty state

attekei opened the pull request:
Addresses #5061 and (1) and (2) of #4809.

attekei committed:
Say "You don't have any saved questions yet" if no questions

mazameli reviewed (changes requested):
It doesn't appear to be quite right for non-admins. Even with no collections, they're seeing "Questions that aren't in a collection will be shown here." That message should only be present if at least one collection is visible to them.

attekei commented:
True, I applied this new text only for admins when there are no questions/collections and left other logic intact. I'll take a look if I could simplify the logic and show the text also to non-admins when questions/collections don't exist.

attekei committed:
Streamline the QuestionIndex component display logic

attekei committed:
Make QuestionIndex testable and add Jest tests

attekei commented:
Now this addresses (1) and (2) of #4809 as well. I found `<QuestionIndex>` a good target for testing and tried to refactor parts of QuestionIndex so that they are very easily testable. I'm interested to hear comments about has QuestionIndex became easy to reason about and do you think that the tests in `QuestionIndex.spec.js` are useful.

tlrobinson commented on frontend/src/metabase/questions/containers/QuestionIndex.jsx:
```diff
@@ -49,81 +85,56 @@ export default class QuestionIndex extends Component {
-            <Button primary>Create a collection</Button>
-        </Link>
-    </div>
+const mapStateToProps = (state, props) => ({
```
In most of the app we keep the `mapDispatchToProps` and `mapDispatchToProps` at the top. I think it's nice to see a component is `@connect`ed up front. Is there a reason you moved it?

attekei replied:
Just because they are now closer to `connect`, but I think that your argument wins in this case. Will move them back up.

mazameli commented:
Nice progress. Two small fixes:

- If there is a least one collection, the heading above collections should read "Collections of Questions," not "Saved Questions." The "Saved Questions" heading is reserved for the state when there are no collections.

![screen shot 2017-05-12 at 11 09 07 am](https://cloud.githubusercontent.com/assets/2223916/26011216/0fc66cc0-3705-11e7-9fb8-8df82372a2b8.png)

- How hard do you think it would be to "upgrade" the empty state for questions to match what we've done for the new dashboards list? Something like this:

![admin](https://cloud.githubusercontent.com/assets/2223916/26011618/b18301a8-3706-11e7-83f8-547612c8e5e5.png)

![non-admin](https://cloud.githubusercontent.com/assets/2223916/26011557/77541044-3706-11e7-9484-b6639e4027de.png)

attekei committed:
Add questions illustration

attekei committed:
QuestionIndex cleanup, fix header glitch

attekei committed:
New beautiful questions list empty state

attekei committed:
Address Enzyme quirks in tests

kdoh commented on frontend/src/metabase/questions/containers/QuestionIndex.jsx:
```diff
@@ -20,110 +20,129 @@ import { getAllCollections, getAllEntities } from "../selectors";
+            <div className="mt1">
+                Collections help you organize your questions and allow you to decide who gets to see what.
+                {" "}
+                <a href="http://www.metabase.com/docs/latest/administration-guide/06-collections.html" target="_blank">
```
Should we be using `/latest` here or should we be using the current version number?

attekei replied:
Seems that the version number tag is used only in the links of setup wizard, elsewhere `/latest` is being used.

tlrobinson replied:
In the future it might be good to add a `Urls.docs` method in `lib/urls`

kdoh commented on frontend/src/metabase/questions/containers/QuestionIndex.jsx:
```diff
@@ -20,110 +20,129 @@ import { getAllCollections, getAllEntities } from "../selectors";
+                </a>
+            </div>
+        </div>
+        <Link to="/collections/create">
```
Curious if we're going all in on the `lib/urls` approach for URLS and if it would be appropriate to extract this there?

attekei replied:
I'm not sure, I have thought that the methods in `lib/urls` are primarily helpers for constructing more complicated urls.

kdoh reviewed (approved):
Left a couple minor code questions but nothing blocking from that side.

mazameli commented:
Comparing the current design for the empty state for the Questions page to the empty state for Dashboards, seems like maybe we should actually leave the `Saved Questions` heading present, even in the empty state. We should probably also make sure the archive icon is the same size on both pages (I tend to prefer the slightly larger version that the dashboard page is using).

They also have a different background color now because we changed the background color of the dashboard list page. This is a trickier problem because I don't think we want to make the questions list page have that background; at least not with the rest of the current visual design it has. A simple fix would be to make just the _empty_ state of the question list match, or to make the empty state of the _dashboard_ list be white. 

@kdoh I know you were doing some thinking on this subject recently; any opinions?

<img width="711" alt="screen shot 2017-05-15 at 11 10 45 am" src="https://cloud.githubusercontent.com/assets/2223916/26072204/9950c13c-395f-11e7-9ede-131a2efbe93d.png">

<img width="711" alt="screen shot 2017-05-15 at 11 10 44 am" src="https://cloud.githubusercontent.com/assets/2223916/26072125/525ff9e6-395f-11e7-8656-5a54d7bccf1c.png">

kdoh commented:
@mazameli / @attekei Yeah I think the move for right now would be to just have the questions empty state match the background color we're using on dashboards.

attekei committed:
Use the same shade of grey in both dash and question lists

attekei committed:
Add loading state, fix favorites filter, sync button sizes

attekei committed:
Fix Jest tests

attekei commented:
Changed the questions list background color, synced the button sizes and added a loading state to the questions list (this prevents some blinking while the list is loading). I also found a glitch with the favorites filter (choosing it hided the filter widget if there were no results) so I fixed it too.

Should be ready to merge.
